test/e2e/network/endpointslice.go

```diff
@@ -0,0 +1,428 @@
+					Ports: []v1.ServicePort{{
+						Name:       "http",
+						Port:       80,
+						TargetPort: intstr.FromString("example-name"),
+						Protocol:   v1.ProtocolTCP,
+					}},
+				},
+			})
+
+			svc3 := createServiceReportErr(cs, f.Namespace.Name, &v1.Service{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "example-no-match",
+				},
+				Spec: v1.ServiceSpec{
+					Selector: map[string]string{labelPod3: labelValue},
+					Ports: []v1.ServicePort{{
+						Name:       "example-no-match",
+						Port:       80,
+						TargetPort: intstr.FromInt(8080),
+						Protocol:   v1.ProtocolTCP,
+					}},
+				},
+			})
+
+			pod1 := podClient.Create(&v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "pod1",
+					Labels: map[string]string{
+						labelPod1:     labelValue,
+						labelShared12: labelValue,
+					},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  "container1",
+							Image: imageutils.GetE2EImage(imageutils.Nginx),
+							Ports: []v1.ContainerPort{{
+								Name:          "example-name",
+								ContainerPort: int32(3000),
+							}},
+						},
+					},
+				},
+			})
+
+			pod2 := podClient.Create(&v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "pod2",
+					Labels: map[string]string{
+						labelPod2:     labelValue,
+						labelShared12: labelValue,
+					},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						{
+							Name:  "container1",
+							Image: imageutils.GetE2EImage(imageutils.Nginx),
+							Ports: []v1.ContainerPort{{
+								Name:          "example-name",
+								ContainerPort: int32(3001),
+							}, {
+								Name:          "other-port",
+								ContainerPort: int32(3002),
+							}},
+						},
+					},
+				},
+			})
+
+			pod1 = waitForPodWithIP(cs, f.Namespace.Name, pod1.Name)
+			pod2 = waitForPodWithIP(cs, f.Namespace.Name, pod2.Name)
+
+			ginkgo.By("referencing a single matching pod")
+			expectEndpointsAndSlices(cs, f.Namespace.Name, svc1, []*v1.Pod{pod1}, 1, 1, false)
+
+			ginkgo.By("referencing matching pods with named port")
+			expectEndpointsAndSlices(cs, f.Namespace.Name, svc2, []*v1.Pod{pod1, pod2}, 2, 2, true)
+
+			ginkgo.By("creating empty endpoints and endpointslices for no matching pods")
+			expectEndpointsAndSlices(cs, f.Namespace.Name, svc3, []*v1.Pod{}, 0, 1, false)
+
+		})
+	})
+})
+
+// expectEndpointsAndSlices verifies that Endpoints and EndpointSlices exist for
+// a given Service and Namespace with the appropriate attributes set. This is a
+// relatively complex function as the order of attributes or resources is not
+// necessarily consistent. It is used as a helper function for the tests above
+// and takes some shortcuts with the assumption that those test cases will be
+// the only caller of this function.
+func expectEndpointsAndSlices(cs clientset.Interface, ns string, svc *v1.Service, pods []*v1.Pod, numSubsets, numSlices int, namedPort bool) {
```

Comment:
Consider having a overarching poll, instead of waiting per object.
This may be more fault tolerant. 
```
wait.Poll(
     pod is ready?
     getEndpoints
     getEndpointSlice
     getMatchingEndpoint
     parse 
)
    

```